include/irr/asset/ITexturePacker.h

```diff
@@ -0,0 +1,483 @@
+    }
+
+    virtual ~ITexturePacker() = default;
+
+    page_tab_offset_t alloc(const IImage* _img, const IImage::SSubresourceRange& _subres)
+    {
+        uint32_t szAndAlignment = computeSquareSz(_img, _subres);
+        szAndAlignment *= szAndAlignment;
+
+        uint32_t addr{};
+        core::address_allocator_traits<pg_tab_addr_alctr_t>::multi_alloc_addr(m_pgTabAddrAlctr, 1u, &addr, &szAndAlignment, &szAndAlignment, nullptr);
+        return (addr==pg_tab_addr_alctr_t::invalid_address) ? 
+            page_tab_offset_invalid() :
+            page_tab_offset_t(core::morton2d_decode_x(addr), core::morton2d_decode_y(addr));
+    }
+    virtual void free(page_tab_offset_t _addr, const IImage* _img, const IImage::SSubresourceRange& _subres)
+    {
+        if ((_addr==page_tab_offset_invalid()).all())
+            return;
+
+        uint32_t sz = computeSquareSz(_img, _subres);
+        sz *= sz;
+        const uint32_t addr = core::morton2d_encode(_addr.x, _addr.y);
+
+        core::address_allocator_traits<pg_tab_addr_alctr_t>::multi_free_addr(m_pgTabAddrAlctr, 1u, &addr, &sz);
+    }
+
+private:
+    uint32_t computeSquareSz(const IImage* _img, const ICPUImage::SSubresourceRange& _subres)
+    {
+        auto extent = _img->getCreationParameters().extent;
+
+        const uint32_t w = std::max(extent.width>>_subres.baseMipLevel,1u);
+        const uint32_t h = std::max(extent.height>>_subres.baseMipLevel,1u);
+
+        return core::roundUpToPoT((std::max(w,h)+m_pgSzxy-1u)/m_pgSzxy);
+    }
+};
+
+class ICPUTexturePacker : public core::IReferenceCounted, protected ITexturePacker
+{
+protected:
+    virtual ~ICPUTexturePacker() = default;
+
+public:
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t MAX_PHYSICAL_PAGE_SIZE_LOG2 = 9u;
+    _IRR_STATIC_INLINE_CONSTEXPR uint32_t MAX_PHYSICAL_PAGE_SIZE = 1u<<MAX_PHYSICAL_PAGE_SIZE_LOG2;
+
+    uint32_t maxAllocatableTextureSz() const { return m_pgSzxy<<m_pageTable->getCreationParameters().mipLevels; }
+
+    uint32_t physAddrTexLayerSz() const
+    {
+        return m_tilesPerDim * (2u*TILE_PADDING+m_pgSzxy);
+    }
+    //! @returns texel-wise offset of physical page
+    core::vector3du32_SIMD pageCoords(SPhysPgOffset _txoffset, uint32_t _pgSz) const
+    {
+        core::vector3du32_SIMD coords(physPgOffset_x(_txoffset), physPgOffset_y(_txoffset), 0u);
+        coords *= (_pgSz + 2u*TILE_PADDING);
+        coords += TILE_PADDING;
+        coords.z = physPgOffset_layer(_txoffset);
+        return coords;
+    }
+    struct SMiptailPacker
+    {
+        struct rect
+        {
+            int x, y, mx, my;
+        };
+        static bool computeMiptailOffsets(rect* res, int log2SIZE, int padding=9);
+    };
+
+    ICPUTexturePacker(E_FORMAT_CLASS _fclass, E_FORMAT _format, uint32_t _pgTabSzxy, uint32_t _maxMipLevels, uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u, uint32_t _tilePad = 9u/*max_aniso/2+1*/) :
```

Comment:
need sane default parameters for the `_pgTabSzxy`, `_tilesPerDim`

Reply:
`_maxMipLevels` is the number of levels of the page table, not the max levels of the allocated texture.. will be confusing.

Because if we change the tile size then the max mip level will change also, I'd rather have it as the max texture mipmap level (and hence max size)